presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPlanWithEmptyInput.java

```diff
@@ -0,0 +1,329 @@
+                return convertToEmptyValuesNode(node);
+            }
+            else if (nonEmptyChildIndex.size() == 1) {
+                this.planChanged = true;
+                int index = nonEmptyChildIndex.get(0);
+                Assignments.Builder builder = Assignments.builder();
+                builder.putAll(node.getVariableMapping().entrySet().stream().collect(toImmutableMap(entry -> entry.getKey(), entry -> entry.getValue().get(index))));
+                return new ProjectNode(node.getSourceLocation(), idAllocator.getNextId(), rewrittenChildren.get(index), builder.build(), LOCAL);
+            }
+            return node.replaceChildren(rewrittenChildren);
+        }
+
+        @Override
+        public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            PlanNode rewrittenFilterSource = context.rewrite(node.getFilteringSource());
+            if (isEmptyNode(rewrittenSource)) {
+                return convertToEmptyValuesNode(node);
+            }
+            else if (isEmptyNode(rewrittenFilterSource)) {
+                this.planChanged = true;
+                Assignments.Builder builder = Assignments.builder();
+                builder.putAll(node.getOutputVariables().stream().collect(toImmutableMap(identity(), x -> x.equals(node.getSemiJoinOutput()) ? constant(false, BOOLEAN) : x)));
+                return new ProjectNode(node.getSourceLocation(), idAllocator.getNextId(), rewrittenSource, builder.build(), LOCAL);
+            }
+            return node.replaceChildren(ImmutableList.of(rewrittenSource, rewrittenFilterSource));
+        }
+
+        @Override
+        public PlanNode visitWindow(WindowNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitSample(SampleNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitOffset(OffsetNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitSort(SortNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            return convertToEmptyNodeIfInputEmpty(node, context);
+        }
+
+        @Override
+        public PlanNode visitLimit(LimitNode node, RewriteContext<Void> context)
```

Comment:
Btw we can always remove Limit from Limit(Values()) by dropping extra rows. We don't have to implement that in this PR(or this rule) however

Reply:
Yeah, I meant something else. Limit 10 -> Values(100 rows) -> Values(10) rows

But that is not related to this PR, and probably not that useful as well